keps/sig-scheduling/5055-dra-device-taints-and-tolerations/README.md

```diff
@@ -0,0 +1,787 @@
+
+A device can be identified by its names (`<driver name>/<pool name>/<device
+name>`) and/or by its attributes (for example, a unique ID). It was a conscious
+decision for core DRA to not require that the name is tied to one particular
+hardware instance to support hot-swapping. Admins might favor using the names
+whereas health monitoring might prefer to be specific and use a vendor-defined
+unique ID. Both are supported, which creates additional complexity.
+
+Without a kubectl extension similar to `kubectl taint nodes`, the user
+experience for admins will be a bit challenging. They need to decide how to
+identify the device (by name or with a CEL expression), manually create a
+ResourceSlicePatch with a unique name, then remember to remove that
+ResourceSlicePatch again. For beta, support in `kubectl` for common
+operations may be needed.
+
+## Design Details
+
+The feature is following the approach and APIs taken for node taints and
+applies them to devices. A new controller watches tainted devices and deletes
+pods using them unless they tolerate the device taint, similar to the
+[taint-eviction-controller](https://github.com/kubernetes/kubernetes/blob/32130691a4cb8a1034b999341c40e48d197f5465/pkg/controller/tainteviction/taint_eviction.go#L81-L83). A pod which is running or has finalizers will not get removed
+immediately. Instead, the `DeletionTimestamp` gets set. That's okay for
+the purpose of this KEP:
+- The kubelet will stop any running containers and mark the pod as completed.
+- The ResourceClaim controller will remove such a completed pod from the claim's
+  `ReservedFor` and deallocate the claim once it has no consumers.
+
+Taints are cumulative as long as the key and effect pairs are different:
+- Taints defined by an admin in a ResourceSlicePatch get added to the
+  set of taints defined by the DRA driver in a ResourceSlice.
+- Taints with the same key and effect get overwritten, using the same
+  precedence as for attributes.
+
+This merging will be implemented by the same code that also
+overrides device attributes.
+
+To ensure consistency among all pods sharing a ResourceClaim, the toleration
+for taints gets added to the request in a ResourceClaim, not the pod. This also
+avoids conflicts like one pod tolerating a taint for scheduling and some other
+pod not tolerating that.
+
+Device and node taints are applied independently. A node taint applies to all
+pods on a node, whereas a device taint affects claim allocation and only those
+pods using the claim.
+
+### API
+
+The ResourceSlice content gets extended:
+
+```Go
+// BasicDevice defines one device instance.
+type BasicDevice struct {
+    ...
+
+    // If specified, the device's taints.
+    //
+    // The maximum number of taints is 8.
+    //
+    // This is an alpha field and requires enabling the DRADeviceTaints
+    // feature gate.
+    //
+    // +optional
+    // +listType=atomic
+    // +featureGate=DRADeviceTaints
+    Taints []DeviceTaint
```

Comment:
Do we have any API markers here to enforce 8 limit?

We can add kubebuilder validations for out-of-tree controllers..

Reply:
> Do we have any API markers here to enforce 8 limit?

I think such markers exist for generating CRDs, but they wouldn't have any effect in-tree.

> Also, why 8?

As usual, a more or less arbitrary choice. For additional discussion, see https://github.com/kubernetes/enhancements/pull/5034#discussion_r1937631658. Because ResourceSlice is size-constrained, I made it a bit smaller than suggested there.